optaplanner-examples/src/main/java/org/optaplanner/examples/scrabble/optional/score/ScrabbleConstraintProvider.java

```diff
@@ -0,0 +1,69 @@
+package org.optaplanner.examples.scrabble.optional.score;
+
+import org.optaplanner.core.api.score.buildin.hardmediumsoft.HardMediumSoftScore;
+import org.optaplanner.core.api.score.stream.Constraint;
+import org.optaplanner.core.api.score.stream.ConstraintCollectors;
+import org.optaplanner.core.api.score.stream.ConstraintFactory;
+import org.optaplanner.core.api.score.stream.ConstraintProvider;
+import org.optaplanner.core.impl.score.stream.bi.FilteringBiJoiner;
+import org.optaplanner.examples.scrabble.domain.ScrabbleWordDirection;
+import org.optaplanner.examples.scrabble.domain.ScrabbleCell;
+import org.optaplanner.examples.scrabble.domain.ScrabbleWordAssignment;
+
+public class ScrabbleConstraintProvider implements ConstraintProvider {
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[]{
+                                characterConflict(constraintFactory),
+                                noParallelHorizontalNeighbours(constraintFactory),
+                                noParallelVerticalNeighbours(constraintFactory),
+                                outOfGrid(constraintFactory),
+                                maximizeMergesPerWord(constraintFactory),
+                                pullToCenter(constraintFactory)
+        };
+    }
+
+    private Constraint characterConflict(ConstraintFactory cf) {
+        return cf.from(ScrabbleCell.class)
+                 .filter(sc -> sc.getCharacterSet().size() >= 2)
+                 .penalize("Character confict", HardMediumSoftScore.ONE_HARD, sc -> 1 - sc.getCharacterSet().size());
```

Comment:
This is where the score corruption is going to be. DRL does this:

    scoreHolder.addHardConstraintMatch(kcontext, 1 - $size);

So the number will be 0 or less. But that is because in DRL, you need to input *a negative number* to penalize.
In CS, you have the `penalize()` call, and that expects *a positive number*. 

Therefore I think that if you invert the logic of the match weigher here, this score corruption will go away.

Reply:
Thanks for that; (although I still find it odd FAST_ASSERT didn't catch this) Suggestion for Constrant Stream: output a warning to the console whenever it see a penalize/reward with a negative number (if possible)

Reply:
There was another one: `maximizeMergesPerWord` was suppose to be a reward, not a penalty. But I got it working in FULL_ASSERT mode now.

Reply:
Please make it run like that for a couple minutes, sometimes the corruptions are hiding in less likely corners of the state space.

Wrt. the warning - I'll think about it. We generally don't do warnings in Optaplanner, but I'll consider if fail-fast would be a good idea here.

Reply:
Error is probably better than a warning so users will immediately know something wrong instead of having to look at the logs to see why it not working correctly. Although that raises the question if someone has a constraint that can be positive or negative (Drools can do that since its reward function is also its penalize function).

Reply:
I ran it for 30 minutes; no Score Corruption was detected on FULL_ASSERT (manually terminated on Local Search step 1677)